In [1]:
from google.colab import drive
drive.mount('/content/gdrive/')


Mounted at /content/gdrive/


In [2]:
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns

##데이터 불러오기
가설 [특정 지역의 고객은 지역 내 인기 카테고리에서 주로 구매할 것이다] 를 검증하기 위한 데이터셋

In [3]:
customers=pd.read_csv('/content/gdrive/MyDrive/MainQuest04/olist_customers_dataset.csv')

In [4]:
order_items=pd.read_csv('/content/gdrive/MyDrive/MainQuest04/olist_order_items_dataset.csv')

In [5]:
orders=pd.read_csv('/content/gdrive/MyDrive/MainQuest04/olist_orders_dataset.csv')

In [6]:
products=pd.read_csv('/content/gdrive/MyDrive/MainQuest04/olist_products_dataset.csv')

In [7]:
category=pd.read_csv('/content/gdrive/MyDrive/MainQuest04/product_category_name_translation.csv')

##카테고리 이름 영어로 변경
(제품 카테고리명이 포르투갈어로 되어 있음, 영어로 번역된 데이터셋으로 대체)

In [8]:
products = products.merge(category, left_on='product_category_name',right_on='product_category_name', how='left')

In [9]:
products = products.drop(['product_category_name'], axis=1)

##병합
1.주문 테이블 + 주문 상품 테이블

In [10]:
order_products_merged=order_items.merge(products, on='product_id',how='left')

In [11]:
order_products_merged

order_id  order_item_id  \
0       00010242fe8c5a6d1ba2dd792cb16214              1   
1       00018f77f2f0320c557190d7a144bdd3              1   
2       000229ec398224ef6ca0657da4fc703e              1   
3       00024acbcdf0a6daa1e931b038114c75              1   
4       00042b26cf59d7ce69dfabb4e55b4fd9              1   
...                                  ...            ...   
112645  fffc94f6ce00a00581880bf54a75a037              1   
112646  fffcd46ef2263f404302a634eb57f7eb              1   
112647  fffce4705a9662cd70adb13d4a31832d              1   
112648  fffe18544ffabc95dfada21779c9644f              1   
112649  fffe41c64501cc87c801fd61db3f6244              1   

                              product_id                         seller_id  \
0       4244733e06e7ecb4970a6e2683c13e61  48436dade18ac8b2bce089ec2a041202   
1       e5f2d52b802189ee658865ca93d83a8f  dd7ddc04e1b6c2c614352b383efe2d36   
2       c777355d18b72b67abbeef9df44fd0fd  5b51032eddd242adc84c38acab88f23d   
3       7634da152a4610f1595efa32f14722fc  9d7a1d34a5052409006425275ba1c2b4   
4       ac6c3623068f30de03045865e4e10089  df560393f3a51e74553ab94004ba5c87   
...                                  ...                               ...   
112645  4aa6014eceb682077f9dc4bffebc05b0  b8bc237ba3788b23da09c0f1f3a3288c   
112646  32e07fd915822b0765e448c4dd74c828  f3c38ab652836d21de61fb8314b69182   
112647  72a30483855e2eafc67aee5dc2560482  c3cfdc648177fdbbbb35635a37472c53   
112648  9c422a519119dcad7575db5af1ba540e  2b3e4a2a3ea8e01938cabda2a3e5cc79   
112649  350688d9dc1e75ff97be326363655e01  f7ccf836d21b2fb1de37564105216cc1   

        shipping_limit_date   price  freight_value  product_name_lenght  \
0       2017-09-19 09:45:35   58.90          13.29                 58.0   
1       2017-05-03 11:05:13  239.90          19.93                 56.0   
2       2018-01-18 14:48:30  199.00          17.87                 59.0   
3       2018-08-15 10:10:18   12.99          12.79                 42.0   
4       2017-02-13 13:57:51  199.90          18.14                 59.0   
...                     ...     ...            ...                  ...   
112645  2018-05-02 04:11:01  299.99          43.41                 43.0   
112646  2018-07-20 04:31:48  350.00          36.53                 31.0   
112647  2017-10-30 17:14:25   99.90          16.95                 43.0   
112648  2017-08-21 00:04:32   55.99           8.72                 56.0   
112649  2018-06-12 17:10:13   43.00          12.79                 47.0   

        product_description_lenght  product_photos_qty  product_weight_g  \
0                            598.0                 4.0             650.0   
1                            239.0                 2.0           30000.0   
2                            695.0                 2.0            3050.0   
3                            480.0                 1.0             200.0   
4                            409.0                 1.0            3750.0   
...                            ...                 ...               ...   
112645                      1002.0                 3.0           10150.0   
112646                       232.0                 1.0            8950.0   
112647                       869.0                 1.0             967.0   
112648                      1306.0                 1.0             100.0   
112649                       511.0                 1.0             600.0   

        product_length_cm  product_height_cm  product_width_cm  \
0                    28.0                9.0              14.0   
1                    50.0               30.0              40.0   
2                    33.0               13.0              33.0   
3                    16.0               10.0              15.0   
4                    35.0               40.0              30.0   
...                   ...                ...               ...   
112645               89.0               15.0              40.0   
112646               45.0               26.0

2.주문 테이블 + 고객 테이블

In [12]:
customers=customers.merge(orders,on='customer_id',how='left')

In [13]:
customers

customer_id                customer_unique_id  \
0      06b8999e2fba1a1fbc88172c00ba8bc7  861eff4711a542e4b93843c6dd7febb0   
1      18955e83d337fd6b2def6b18a428ac77  290c77bc529b7ac935b93aa66c333dc3   
2      4e7b3e00288586ebd08712fdd0374a03  060e732b5b29e8181a18229c7b0b2b5e   
3      b2b6027bc5c5109e529d4dc6358b12c3  259dac757896d24d7702b9acbbff3f3c   
4      4f2d8ab171c80ec8364f7c12e35b23ad  345ecd01c38d18a9036ed96c73b8d066   
...                                 ...                               ...   
99436  17ddf5dd5d51696bb3d7c6291687be6f  1a29b476fee25c95fbafc67c5ac95cf8   
99437  e7b71a9017aa05c9a7fd292d714858e8  d52a67c98be1cf6a5c84435bd38d095d   
99438  5e28dfe12db7fb50a4b2f691faecea5e  e9f50caf99f032f0bf3c55141f019d99   
99439  56b18e2166679b8a959d72dd06da27f9  73c2643a0a458b49f58cea58833b192e   
99440  274fa6071e5e17fe303b9748641082c8  84732c5050c01db9b23e19ba39899398   

       customer_zip_code_prefix          customer_city customer_state  \
0                         14409                 franca             SP   
1                          9790  sao bernardo do campo             SP   
2                          1151              sao paulo             SP   
3                          8775        mogi das cruzes             SP   
4                         13056               campinas             SP   
...                         ...                    ...            ...   
99436                      3937              sao paulo             SP   
99437                      6764        taboao da serra             SP   
99438                     60115              fortaleza             CE   
99439                     92120                 canoas             RS   
99440                      6703                  cotia             SP   

                               order_id order_status order_purchase_timestamp  \
0      00e7ee1b050b8499577073aeb2a297a1    delivered      2017-05-16 15:05:35   
1      29150127e6685892b6eab3eec79f59c7    delivered      2018-01-12 20:48:24   
2      b2059ed67ce144a36e2aa97d2c9e9ad2    delivered      2018-05-19 16:07:45   
3      951670f92359f4fe4a63112aa7306eba    delivered      2018-03-13 16:06:38   
4      6b7d50bd145f6fc7f33cebabd7e49d0f    delivered      2018-07-29 09:51:30   
...                                 ...          ...                      ...   
99436  6760e20addcf0121e9d58f2f1ff14298    delivered      2018-04-07 15:48:17   
99437  9ec0c8947d973db4f4e8dcf1fbfa8f1b    delivered      2018-04-04 08:20:22   
99438  fed4434add09a6f332ea398efd656a5c    delivered      2018-04-08 20:11:50   
99439  e31ec91cea1ecf97797787471f98a8c2    delivered      2017-11-03 21:08:33   
99440  28db69209a75e59f20ccbb5c36a20b90    delivered      2017-12-19 14:27:23   

         order_approved_at order_delivered_carrier_date  \
0      2017-05-16 15:22:12          2017-05-23 10:47:57   
1      2018-01-12 20:58:32          2018-01-15 17:14:59   
2      2018-05-20 16:19:10          2018-06-11 14:31:00   
3      2018-03-13 17:29:19          2018-03-27 23:22:42   
4      2018-07-29 10:10:09          2018-07-30 15:16:00   
...                    ...                          ...   
99436  2018-04-07 16:08:45          2018-04-11 02:08:36   
99437  2018-04-04 08:35:12          2018-04-05 18:42:35   
99438  2018-04-08 20:30:03          2018-04-09 17:52:17   
99439  2017-11-03 21:31:20          2017-11-06 18:24:41   
99440  2017-12-19 18:50:39          2017-12-21 19:17:21   

      order_delivered_customer_date order_estimated_delivery_date  
0               2017-05-25 10:35:35           2017-06-05 00:00:00  
1               2018-01-29 12:41:19           2018-02-06 00:00:00  
2               2018-06-14 17:58:51           2018-06-13 00:00:00  
3               2018-03-28 16:04:25           2018-04-10 00:00:00  
4               2018-08-09 20:55:48           2018-08-15 00:00:00  
...                             ...                           ...  
99436           2018-04-13 20:06:37           2018-04-25 00:00:00  


3. 1과2를 병합

(고객별 주문 내역과 주문 상품이 포함된 테이블 생성)



In [14]:
customers_orders=customers.merge(order_products_merged,on='order_id',how='left')

In [15]:
customers_orders

customer_id                customer_unique_id  \
0       06b8999e2fba1a1fbc88172c00ba8bc7  861eff4711a542e4b93843c6dd7febb0   
1       18955e83d337fd6b2def6b18a428ac77  290c77bc529b7ac935b93aa66c333dc3   
2       4e7b3e00288586ebd08712fdd0374a03  060e732b5b29e8181a18229c7b0b2b5e   
3       b2b6027bc5c5109e529d4dc6358b12c3  259dac757896d24d7702b9acbbff3f3c   
4       4f2d8ab171c80ec8364f7c12e35b23ad  345ecd01c38d18a9036ed96c73b8d066   
...                                  ...                               ...   
113420  17ddf5dd5d51696bb3d7c6291687be6f  1a29b476fee25c95fbafc67c5ac95cf8   
113421  e7b71a9017aa05c9a7fd292d714858e8  d52a67c98be1cf6a5c84435bd38d095d   
113422  5e28dfe12db7fb50a4b2f691faecea5e  e9f50caf99f032f0bf3c55141f019d99   
113423  56b18e2166679b8a959d72dd06da27f9  73c2643a0a458b49f58cea58833b192e   
113424  274fa6071e5e17fe303b9748641082c8  84732c5050c01db9b23e19ba39899398   

        customer_zip_code_prefix          customer_city customer_state  \
0                          14409                 franca             SP   
1                           9790  sao bernardo do campo             SP   
2                           1151              sao paulo             SP   
3                           8775        mogi das cruzes             SP   
4                          13056               campinas             SP   
...                          ...                    ...            ...   
113420                      3937              sao paulo             SP   
113421                      6764        taboao da serra             SP   
113422                     60115              fortaleza             CE   
113423                     92120                 canoas             RS   
113424                      6703                  cotia             SP   

                                order_id order_status  \
0       00e7ee1b050b8499577073aeb2a297a1    delivered   
1       29150127e6685892b6eab3eec79f59c7    delivered   
2       b2059ed67ce144a36e2aa97d2c9e9ad2    delivered   
3       951670f92359f4fe4a63112aa7306eba    delivered   
4       6b7d50bd145f6fc7f33cebabd7e49d0f    delivered   
...                                  ...          ...   
113420  6760e20addcf0121e9d58f2f1ff14298    delivered   
113421  9ec0c8947d973db4f4e8dcf1fbfa8f1b    delivered   
113422  fed4434add09a6f332ea398efd656a5c    delivered   
113423  e31ec91cea1ecf97797787471f98a8c2    delivered   
113424  28db69209a75e59f20ccbb5c36a20b90    delivered   

       order_purchase_timestamp    order_approved_at  \
0           2017-05-16 15:05:35  2017-05-16 15:22:12   
1           2018-01-12 20:48:24  2018-01-12 20:58:32   
2           2018-05-19 16:07:45  2018-05-20 16:19:10   
3           2018-03-13 16:06:38  2018-03-13 17:29:19   
4           2018-07-29 09:51:30  2018-07-29 10:10:09   
...                         ...                  ...   
113420      2018-04-07 15:48:17  2018-04-07 16:08:45   
113421      2018-04-04 08:20:22  2018-04-04 08:35:12   
113422      2018-04-08 20:11:50  2018-04-08 20:30:03   
113423      2017-11-03 21:08:33  2017-11-03 21:31:20   
113424      2017-12-19 14:27:23  2017-12-19 18:50:39   

       order_delivered_carrier_date  ...   price freight_value  \
0               2017-05-23 10:47:57  ...  124.99         21.88   
1               2018-01-15 17:14:59  ...  289.00         46.48   
2               2018-06-11 14:31:00  ...  139.94         17.79   
3               2018-03-27 23:22:42  ...  149.94         23.36   
4               2018-07-30 15:16:00  ...  230.00         22.25   
...                             ...  ...     ...           ...   
113420          2018-04-11 02:08:36  ...   74.90         13.88   
113421          2018-04-05 18:42:35  ...  114.90         14.16   
113422          2018-04-09 17:52:17  ...   37.00         19.04   
113423          2017-11-06 18:24:41  ...  689.00         22.07   
113424          2017-12-21 19:17:21  ...   13.99          7.78   

        product_name_lenght product_description_leng

##필요한 컬럼만 남기기
(가설 검증에 불필요한 상품의 무게,배송 날짜 등의 컬럼 삭제)

In [16]:
# 필요한 컬럼 선택
selected_columns = [
    'order_id',
    'customer_unique_id',
    'customer_zip_code_prefix',
    'customer_city',
    'customer_state',
    'order_item_id',
    'product_id',
    'product_category_name_english'
]

# 새로운 데이터프레임 생성
customers_orders_new= customers_orders[selected_columns]


In [17]:
customers_orders_new

order_id                customer_unique_id  \
0       00e7ee1b050b8499577073aeb2a297a1  861eff4711a542e4b93843c6dd7febb0   
1       29150127e6685892b6eab3eec79f59c7  290c77bc529b7ac935b93aa66c333dc3   
2       b2059ed67ce144a36e2aa97d2c9e9ad2  060e732b5b29e8181a18229c7b0b2b5e   
3       951670f92359f4fe4a63112aa7306eba  259dac757896d24d7702b9acbbff3f3c   
4       6b7d50bd145f6fc7f33cebabd7e49d0f  345ecd01c38d18a9036ed96c73b8d066   
...                                  ...                               ...   
113420  6760e20addcf0121e9d58f2f1ff14298  1a29b476fee25c95fbafc67c5ac95cf8   
113421  9ec0c8947d973db4f4e8dcf1fbfa8f1b  d52a67c98be1cf6a5c84435bd38d095d   
113422  fed4434add09a6f332ea398efd656a5c  e9f50caf99f032f0bf3c55141f019d99   
113423  e31ec91cea1ecf97797787471f98a8c2  73c2643a0a458b49f58cea58833b192e   
113424  28db69209a75e59f20ccbb5c36a20b90  84732c5050c01db9b23e19ba39899398   

        customer_zip_code_prefix          customer_city customer_state  \
0                          14409                 franca             SP   
1                           9790  sao bernardo do campo             SP   
2                           1151              sao paulo             SP   
3                           8775        mogi das cruzes             SP   
4                          13056               campinas             SP   
...                          ...                    ...            ...   
113420                      3937              sao paulo             SP   
113421                      6764        taboao da serra             SP   
113422                     60115              fortaleza             CE   
113423                     92120                 canoas             RS   
113424                      6703                  cotia             SP   

        order_item_id                        product_id  \
0                 1.0  a9516a079e37a9c9c36b9b78b10169e8   
1                 1.0  4aa6014eceb682077f9dc4bffebc05b0   
2                 1.0  bd07b66896d6f1494f5b86251848ced7   
3                 1.0  a5647c44af977b148e0a3a4751a09e2e   
4                 1.0  9391a573abe00141c56e38d84d7d5b3b   
...               ...                               ...   
113420            1.0  ccb4503d9d43d245d3b295d0544f988b   
113421            1.0  9ede6b0570a75a4b9de4f383329f99ee   
113422            1.0  7a5d2e1e131a860ae7d18f6fffa9d689   
113423            1.0  f819f0c84a64f02d3a5606ca95edd272   
113424            1.0  017692475c1c954ff597feda05131d73   

       product_category_name_english  
0                   office_furniture  
1                         housewares  
2                   office_furniture  
3                   office_furniture  
4                       home_confort  
...                              ...  
113420        books_general_interest  
113421                sports_leisure  
113422                 health_beauty  
113423                 watches_gifts  
113424                     perfumery  

[113425 rows x 8 columns]

In [39]:
# 결측치의 총 갯수 확인하기
customers_orders_new.isna().sum()

order_id                         0
customer_unique_id               0
customer_zip_code_prefix         0
customer_city                    0
customer_state                   0
order_item_id                    0
product_id                       0
product_category_name_english    0
region                           0
dtype: int64

결측치 날리기
*  order_item, product_id가 널값인것이 category널값이기도함

*   따라서 결측치 모두 drop





In [19]:
customers_orders_new=customers_orders_new.dropna()

In [20]:
customers_orders_new.info()

<class 'pandas.core.frame.DataFrame'>
Index: 111023 entries, 0 to 113424
Data columns (total 8 columns):
 #   Column                         Non-Null Count   Dtype  
---  ------                         --------------   -----  
 0   order_id                       111023 non-null  object 
 1   customer_unique_id             111023 non-null  object 
 2   customer_zip_code_prefix       111023 non-null  int64  
 3   customer_city                  111023 non-null  object 
 4   customer_state                 111023 non-null  object 
 5   order_item_id                  111023 non-null  float64
 6   product_id                     111023 non-null  object 
 7   product_category_name_english  111023 non-null  object 
dtypes: float64(1), int64(1), object(6)
memory usage: 7.6+ MB


In [21]:
customers_orders_new['customer_state'].unique()

array(['SP', 'SC', 'PR', 'MG', 'RJ', 'RS', 'PA', 'GO', 'ES', 'BA', 'MA',
       'MS', 'CE', 'DF', 'RN', 'PE', 'MT', 'AM', 'AP', 'AL', 'RO', 'PB',
       'TO', 'PI', 'AC', 'SE', 'RR'], dtype=object)

state의 수가 27개로 너무 많아서,  

*   북부 (North)	AC,AM, AP, PA, RO, RR, TO
*   북동부 (Northeast)	AL, BA, CE, MA, PB, PE, PI, RN, SE
*   중서부 (Central-West)	DF, GO, MT, MS
*   남동부 (Southeast)	ES, MG, RJ, SP
*   남부 (South)	PR, RS, SC

지역별로 추려서 5개의 그룹으로 그룹할 예정






In [23]:
# 주(State)별 지역 그룹화
region_mapping = {
    "North": ["AC", "AM", "AP", "PA", "RO", "RR", "TO"],
    "Northeast": ["AL", "BA", "CE", "MA", "PB", "PE", "PI", "RN", "SE"],
    "Central-West": ["DF", "GO", "MT", "MS"],
    "Southeast": ["ES", "MG", "RJ", "SP"],
    "South": ["PR", "RS", "SC"]
}

# 데이터프레임에 지역 추가
customers_orders_new["region"] = customers_orders_new["customer_state"].map(lambda x: next((k for k, v in region_mapping.items() if x in v), "Unknown"))

# 1. 지역(Region)별 인기 제품 분석
region_popular_products = customers_orders_new.groupby(['region', 'product_category_name_english']).size().reset_index(name='purchase_count')

# 각 지역에서 가장 많이 판매된 제품 카테고리 찾기
top_products_per_region = region_popular_products.sort_values(['region', 'purchase_count'], ascending=[True, False]).groupby('region').first().reset_index()

# 2. 특정 지역 고객의 구매 패턴 분석
# 고객이 구매한 제품과 지역별 인기 제품 매칭
customer_purchases = customers_orders_new.merge(top_products_per_region, on='region', suffixes=('_customer', '_region'))

# 고객이 자신의 지역에서 인기 있는 제품을 구매했는지 확인
customer_purchases['matches_region_popularity'] = customer_purchases['product_category_name_english_customer'] == customer_purchases['product_category_name_english_region']

# 지역(region)별 인기 제품을 구매한 고객 비율 계산
region_popularity_match_rate = customer_purchases.groupby('region')['matches_region_popularity'].mean().reset_index()

# 3. 지역(region)별 주문 통계량 계산
region_stats = customers_orders_new.groupby('region').agg(
    total_orders_per_region=('order_id', 'count'),  # 총 주문 개수
    unique_customers_per_region=('customer_unique_id', 'nunique'),  # 고유 고객 수
    avg_order_items_per_region=('order_item_id', 'mean'),  # 평균 주문 아이템 개수
    avg_product_category_diversity=('product_category_name_english', 'nunique')  # 제품 카테고리 다양성
).reset_index()

# 4. 기존 matches_region_popularity 데이터와 결합
region_correlation_data = region_popularity_match_rate.merge(region_stats, on='region', how='left')

# 5. 상관관계 계산
correlation_matrix = region_correlation_data.drop(columns=['region']).corr()


<ipython-input-23-9d4a1c10c33b>:11: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  customers_orders_new["region"] = customers_orders_new["customer_state"].map(lambda x: next((k for k, v in region_mapping.items() if x in v), "Unknown"))


In [25]:
top_products_per_region
#각 지역에서 가장 많이 판매된 카테고리

region product_category_name_english  purchase_count
0  Central-West                 health_beauty             630
1         North                 health_beauty             212
2     Northeast                 health_beauty            1139
3         South               furniture_decor            1435
4     Southeast                bed_bath_table            8435

In [36]:
customer_purchases['matches_region_popularity'].sum()
#지역별 인기제품을 구매한 고객의 수

11851

In [38]:
region_popularity_match_rate*100
#지역별 인기제품을 구매한 고객의 비율(%)

region  \
0  Central-WestCentral-WestCentral-WestCentral-We...   
1  NorthNorthNorthNorthNorthNorthNorthNorthNorthN...   
2  NortheastNortheastNortheastNortheastNortheastN...   
3  SouthSouthSouthSouthSouthSouthSouthSouthSouthS...   
4  SoutheastSoutheastSoutheastSoutheastSoutheastS...   

   matches_region_popularity  
0                   9.684858  
1                  10.458806  
2                  11.094876  
3                   9.020619  
4                  11.052583

In [24]:
correlation_matrix
#상관관계

matches_region_popularity  \
matches_region_popularity                        1.000000   
total_orders_per_region                          0.414763   
unique_customers_per_region                      0.418877   
avg_order_items_per_region                      -0.661890   
avg_product_category_diversity                  -0.127980   

                                total_orders_per_region  \
matches_region_popularity                      0.414763   
total_orders_per_region                        1.000000   
unique_customers_per_region                    0.999985   
avg_order_items_per_region                     0.093424   
avg_product_category_diversity                 0.539128   

                                unique_customers_per_region  \
matches_region_popularity                          0.418877   
total_orders_per_region                            0.999985   
unique_customers_per_region                        1.000000   
avg_order_items_per_region                         0.089306   
avg_product_category_diversity                     0.539447   

                                avg_order_items_per_region  \
matches_region_popularity                        -0.661890   
total_orders_per_region                           0.093424   
unique_customers_per_region                       0.089306   
avg_order_items_per_region                        1.000000   
avg_product_category_diversity                    0.385967   

                                avg_product_category_diversity  
matches_region_popularity                            -0.127980  
total_orders_per_region                               0.539128  
unique_customers_per_region                           0.539447  
avg_order_items_per_region                            0.385967  
avg_product_category_diversity                        1.000000

1. matches_region_popularity vs total_orders_per_region (0.415, 양의 상관관계)
→ 주문량이 많은 지역에서 인기 제품을 더 많이 구매하는 경향이 있음.
→ 인기 제품 추천 전략이 효과적인 지역을 찾는 데 유용.

2. matches_region_popularity vs unique_customers_per_region (0.419, 양의 상관관계)
→ 고객 수가 많은 지역일수록 인기 제품을 더 많이 구매하는 경향이 있음.
→ 고객이 많은 지역에서 인기 제품 추천이 효과적일 가능성 큼.

3. matches_region_popularity vs avg_order_items_per_region (-0.662, 강한 음의 상관관계)
→ 한 번에 여러 개의 제품을 구매하는 고객이 많은 지역에서는 인기 제품을 따르지 않는 경향이 강함.
→ 개별 맞춤 추천이 필요한 지역을 찾는 데 유용.

4. matches_region_popularity vs avg_product_category_diversity (-0.128, 약한 상관관계)
카테고리 다양성이 높은 지역에서는 인기 제품을 따르는 경향이 약함.
해석:
하지만 -0.128로 상관관계가 약하므로 영향이 크지 않음.
단, 카테고리 다양성이 높은 고객에게는 개별 맞춤 추천이 필요할 수 있음.

##가설검정 수행하기

가설 검정에 필요한 컬럼만 추리기

1.  order_id : 주문ID
2.   customer_unique_id : 고객 고유 ID
3. region : 고객이 거주하는 지역
3. product_category_name_english_customer : 고객이 주문한 상품 카테고리
4. product_category_name_english_region : 해당 지역의 인기 카테고리
4. matches_region_popularity : 고객이 지역 내 인기 제품을 구매했는지 여부 (True/False)



In [45]:
test_columns=['order_id','customer_unique_id','region','product_category_name_english_customer','product_category_name_english_region','matches_region_popularity']
T_test_df=customer_purchases[test_columns]


In [47]:
from scipy.stats import ttest_1samp


In [52]:
# ✅ 특정 지역의 고객이 지역 내 인기 제품을 선호하는지 검정
# - matches_region_popularity 비율이 50%와 유의미하게 다른지 검증

# 1. 가설 설정
# H0: 특정 지역의 고객은 지역 내 인기 제품을 선호하지 않는다. (비율 = 50%)
# H1: 특정 지역의 고객은 지역 내 인기 제품을 선호한다. (비율 ≠ 50%)

# 2. 유의수준 설정
alpha = 0.05

# 3. 검정통계량 산출
test_mean = 0.50  # 기준 비율 (40%)
t_stat, p_value = ttest_1samp(T_test_df['matches_region_popularity'], test_mean)

# 4. 가설 기각/채택
if p_value < alpha:
    result = "귀무가설 기각 → 특정 지역의 고객은 지역 내 인기 제품을 선호한다."
else:
    result = "귀무가설 채택 → 특정 지역의 고객은 지역 내 인기 제품을 선호하지 않는다."

# 5. 결과 출력
print(f"t-통계량: {t_stat:.4f}")
print(f"p-value: {p_value:.4f}")
print(f"결론: {result}")

t-통계량: -424.3471
p-value: 0.0000
결론: 귀무가설 기각 → 특정 지역의 고객은 지역 내 인기 제품을 선호한다.


- p-value = 0.0000 (< 0.05)
→ 유의수준(α = 0.05)보다 작기 때문에, 귀무가설(H₀) 기각
→ 즉, 특정 지역 고객이 해당 지역에서 인기 제품을 특별히 더 선호한다는 강력한 증거가 있음.

- t-통계량 = -424.3471
→ 매우 낮은 t-값 → 검정 기준(test_mean= 50%)보다 훨씬 낮은 값을 나타냄.
→ 실제 고객들의 인기 제품 구매율이 50%보다 훨씬 낮을 가능성이 큼.
→ 즉, 특정 지역 고객이 인기 제품을 선호하긴 하지만, 예상보다 덜 구매하고 있을 가능성이 있음.